p2p/src/secret_connection.rs

```diff
@@ -357,18 +392,23 @@ where
 {
     // CONTRACT: data smaller than DATA_MAX_SIZE is read atomically.
     fn read(&mut self, data: &mut [u8]) -> io::Result<usize> {
-        if !self.recv_buffer.is_empty() {
-            let n = cmp::min(data.len(), self.recv_buffer.len());
-            data.copy_from_slice(&self.recv_buffer[..n]);
+        let mut recv_state = self
+            .recv_state
+            .lock()
+            .expect("failed to obtain lock on secret connection recv state");
```

Comment:
Should we return error result here instead of panicking? Same for the other uses of `.expect()`.

Reply:
How would we (and should we be able to) recover from such an error?

My understanding is that, when [`Mutex::lock`](https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.lock) fails, it's due to a panic while the lock is held by another thread.

Reply:
Sounds reasonable. Perhaps we can add a comment here that we are panicking because a poisoned mutex is unrecoverable. This is mainly to distinguish it from the other uses of `.expect` in the same file, which IMO should be refactored to return `Result::Err` instead.

Reply:
Fortunately in the new iteration there's no need for mutexes any more :grin: